src/chain/store/index.rs

```diff
@@ -41,12 +47,54 @@ pub enum ResolveNullTipset {
 impl<DB: Blockstore> ChainIndex<DB> {
     pub fn new(db: DB) -> Self {
         let ts_cache = Mutex::new(LruCache::new(DEFAULT_TIPSET_CACHE_SIZE));
-        Self { ts_cache, db }
+        Self {
+            ts_cache,
+            db,
+            counter: AtomicU64::new(0),
+        }
     }
 
     /// Loads a tipset from memory given the tipset keys and cache. Semantically
     /// identical to [`Tipset::load`] but the result is cached.
     pub fn load_tipset(&self, tsk: &TipsetKey) -> Result<Option<Arc<Tipset>>, Error> {
+        self.counter
+            .fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+        // every now and then, check the cache size
+        if self.counter.load(std::sync::atomic::Ordering::Relaxed) % 10000 == 0 {
+            let timer = Instant::now();
+            let total = self
+                .ts_cache
+                .lock()
+                .iter()
+                .map(|(tsk, ts)| {
+                    let tipset_size = tsk.len() * 40;
+                    let ts_size = ts
+                        .block_headers()
+                        .iter()
+                        .map(|bh| {
+                            let raw = bh.clone().into_raw();
+                            raw.get_size() + std::mem::size_of::<RawBlockHeader>()
+                        })
+                        .sum::<usize>();
+                    tipset_size + ts_size
+                })
+                .sum::<usize>();
+            info!(
+                "Tipset cache size: {} bytes ({} MB), {} entries in cache, took {} ms",
+                total,
+                total / (1024 * 1024),
+                self.ts_cache.lock().len(),
+                timer.elapsed().as_millis()
+            );
+
+            metrics::TIPSET_CACHE_SIZE
```

Comment:
It seems to be expensive to always update the metrics, I'd register a `prometheus_client::collector::Collector` instead